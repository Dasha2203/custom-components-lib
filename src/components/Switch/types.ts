import { InputHTMLAttributes } from 'react';

export type SwitchProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  'type'
> & {
  label?: string;
};
