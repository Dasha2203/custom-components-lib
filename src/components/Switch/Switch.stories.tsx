import type { Meta, StoryObj } from '@storybook/react';

import Switch from './Switch';

const meta = {
  title: 'Switch',
  component: Switch,
  tags: ['autodocs'],
} satisfies Meta<typeof Switch>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Off: Story = {
  args: {
    checked: false,
  },
};

export const On: Story = {
  args: {
    checked: true,
  },
};

export const LockedOn: Story = {
  args: {
    checked: true,
    disabled: true,
  },
};

export const LockedOff: Story = {
  args: {
    checked: false,
    disabled: true,
  },
};
