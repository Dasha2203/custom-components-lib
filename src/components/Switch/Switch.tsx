import clsx from 'clsx';
import * as classes from './style.module.scss';
import { SwitchProps } from './types';

const Switch = ({ checked, disabled, ...props }: SwitchProps) => {
  return (
    <label
      className={clsx(
        classes.switch,
        checked && classes.switch_checked,
        disabled && classes.switch_disabled
      )}
    >
      <span className={classes.switch__toggle}>
        <input
          type={'checkbox'}
          className={classes.switch__input}
          disabled={disabled}
          checked={checked}
          {...props}
        />
        <span className={classes.switch__thumb}></span>
      </span>
      <span className={classes.switch__track}></span>
    </label>
  );
};

export default Switch;
