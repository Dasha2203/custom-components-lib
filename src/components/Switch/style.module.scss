.switch {
  position: relative;
  z-index: 0;
  padding: 12px;
  display: inline-flex;
  width: 58px;
  height: 38px;
  overflow: hidden;
  box-sizing: border-box;

  &__toggle {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
    padding: 9px;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    -webkit-tap-highlight-color: transparent;
    background-color: transparent;
    cursor: pointer;
    user-select: none;
    vertical-align: middle;
    text-decoration: none;
    border-radius: 50%;
    transition:
      left 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    color: #fff;
  }

  &__input {
    position: absolute;
    margin: 0;
    top: 0;
    z-index: 1;
    padding: 0;
    height: 100%;
    cursor: inherit;
    opacity: 0;
  }

  &__thumb {
    box-shadow:
      0px 2px 1px -1px rgba(0, 0, 0, 0.2),
      0px 1px 1px 0px rgba(0, 0, 0, 0.14),
      0px 1px 3px 0px rgba(0, 0, 0, 0.12);
    background-color: currentColor;
    width: 20px;
    height: 20px;
    border-radius: 50%;
  }

  &__track {
    height: 100%;
    width: 100%;
    border-radius: 7px;
    z-index: -1;
    transition:
      opacity 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    background-color: #000;
    opacity: 0.38;
  }

  &_checked &__toggle {
    color: #1976d2;
    transform: translateX(20px);
  }

  &_checked &__track {
    background-color: #1976d2;
  }

  &_disabled &__track {
    opacity: 0.12;
  }

  &_disabled &__toggle {
    color: #f5f5f5;
    pointer-events: none;
    cursor: default;
  }

  &.switch_checked.switch_disabled &__toggle {
    color: rgb(167, 202, 237);
  }

  &.switch_checked.switch_disabled &__track {
    background-color: #1976d2;
    opacity: 0.12;
  }
}
