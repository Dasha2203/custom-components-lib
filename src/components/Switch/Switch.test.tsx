import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Switch from './Switch';

describe('Testing Switch', () => {
  test('Should render Switch component', () => {
    render(<Switch />);
    const switchElement = screen.getByRole('checkbox');

    expect(switchElement).toBeInTheDocument();
  });

  test('Should be checked', () => {
    render(<Switch defaultChecked={true} />);
    const switchElement = screen.getByRole('checkbox');

    expect(switchElement).toBeChecked();
  });

  test('Should be unchecked', () => {
    render(<Switch defaultChecked={false} />);
    const switchElement = screen.getByRole('checkbox');

    expect(switchElement).not.toBeChecked();
  });

  test('Should be disabled', async () => {
    render(<Switch checked={false} disabled />);
    const switchElement = screen.getByRole('checkbox');

    expect(switchElement).toBeDisabled();
  });
});

describe('Snapshots of Switch', () => {
  test('Checked Switch', () => {
    const { asFragment } = render(<Switch defaultChecked={true} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Unchecked Switch', () => {
    const { asFragment } = render(<Switch defaultChecked={false} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Disabled checked Switch', () => {
    const { asFragment } = render(<Switch checked={true} disabled />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Disabled unchecked Switch', () => {
    const { asFragment } = render(<Switch checked={false} disabled />);

    expect(asFragment()).toMatchSnapshot();
  });
});
