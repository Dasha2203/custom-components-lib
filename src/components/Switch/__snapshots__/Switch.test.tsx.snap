// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots of Switch Checked Switch 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <span
      class="switch__toggle"
    >
      <input
        checked=""
        class="switch__input"
        type="checkbox"
      />
      <span
        class="switch__thumb"
      />
    </span>
    <span
      class="switch__track"
    />
  </label>
</DocumentFragment>
`;

exports[`Snapshots of Switch Disabled checked Switch 1`] = `
<DocumentFragment>
  <label
    class="switch switch_checked switch_disabled"
  >
    <span
      class="switch__toggle"
    >
      <input
        checked=""
        class="switch__input"
        disabled=""
        type="checkbox"
      />
      <span
        class="switch__thumb"
      />
    </span>
    <span
      class="switch__track"
    />
  </label>
</DocumentFragment>
`;

exports[`Snapshots of Switch Disabled unchecked Switch 1`] = `
<DocumentFragment>
  <label
    class="switch switch_disabled"
  >
    <span
      class="switch__toggle"
    >
      <input
        class="switch__input"
        disabled=""
        type="checkbox"
      />
      <span
        class="switch__thumb"
      />
    </span>
    <span
      class="switch__track"
    />
  </label>
</DocumentFragment>
`;

exports[`Snapshots of Switch Unchecked Switch 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <span
      class="switch__toggle"
    >
      <input
        class="switch__input"
        type="checkbox"
      />
      <span
        class="switch__thumb"
      />
    </span>
    <span
      class="switch__track"
    />
  </label>
</DocumentFragment>
`;
